clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java

```diff
@@ -318,4 +318,9 @@ public AttributeList setAttributes(AttributeList list) {
                                       + ".(whitelist/blacklist) is not a valid regular expression");
         }
     }
+
+    @Override
+    public void contextChange(MetricsContext metricsContext) {
+        this.prefix = metricsContext.metadata().get(MetricsContext.NAMESPACE);
```

Comment:
we shouldn't register any metrics in JMX until the prefix is set. Once it's set we shouldn't modify it, otherwise dynamic reconfiguration of whitelist / blacklists will not work as expected.

Reply:
Check if prefix has been set, if not, set prefix. Make the method synchronized.